vortex-array/src/array/bool/compute/filter.rs

```diff
@@ -1,67 +1,80 @@
-use arrow_buffer::{BooleanBuffer, BooleanBufferBuilder};
-use vortex_error::VortexResult;
+use arrow_buffer::{bit_util, BooleanBuffer, BooleanBufferBuilder};
+use vortex_error::{VortexExpect, VortexResult};
 
 use crate::array::BoolArray;
-use crate::compute::{FilterFn, FilterMask};
+use crate::compute::{FilterFn, FilterIter, FilterMask};
 use crate::{ArrayData, IntoArrayData};
 
 impl FilterFn for BoolArray {
-    fn filter(&self, mask: &FilterMask) -> VortexResult<ArrayData> {
-        filter_select_bool(self, mask).map(|a| a.into_array())
+    fn filter(&self, mask: FilterMask) -> VortexResult<ArrayData> {
+        let validity = self.validity().filter(&mask)?;
+
+        let buffer = match mask.iter()? {
+            FilterIter::Indices(iter) => {
+                filter_indices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+            FilterIter::LazyIndices(iter) => {
+                filter_indices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+            FilterIter::Slices(iter) => {
+                filter_slices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+            FilterIter::LazySlices(iter) => {
+                filter_slices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+        };
+
+        Ok(Self::try_new(buffer, validity)?.into_array())
     }
 }
 
-fn filter_select_bool(arr: &BoolArray, mask: &FilterMask) -> VortexResult<BoolArray> {
-    let validity = arr.validity().filter(mask)?;
-
-    let selection_count = mask.true_count();
-    let out = if selection_count * 2 > arr.len() {
-        filter_select_bool_by_slice(&arr.boolean_buffer(), mask, selection_count)
-    } else {
-        filter_select_bool_by_index(&arr.boolean_buffer(), mask, selection_count)
-    };
-    BoolArray::try_new(out?, validity)
-}
-
-fn filter_select_bool_by_slice(
-    values: &BooleanBuffer,
-    mask: &FilterMask,
-    selection_count: usize,
-) -> VortexResult<BooleanBuffer> {
-    let mut out_buf = BooleanBufferBuilder::new(selection_count);
-    mask.iter_slices()?.for_each(|(start, end)| {
-        out_buf.append_buffer(&values.slice(start, end - start));
-    });
-    Ok(out_buf.finish())
+fn filter_indices(
+    buffer: BooleanBuffer,
+    indices_len: usize,
+    mut indices: impl Iterator<Item = usize>,
+) -> BooleanBuffer {
+    let src = buffer.values().as_ptr();
+    let offset = buffer.offset();
+
+    BooleanBuffer::collect_bool(indices_len, |_idx| {
+        let idx = indices.next().vortex_expect("iterator length must match");
+        unsafe { bit_util::get_bit_raw(src, idx + offset) }
```

Comment:
given that this bypasses bounds checks, do we check this by construction somewhere?

Reply:
Yes, the filter compute function asserts the FilterMask.len == the array.len. Filter mask indices are then derived from this.